spynnaker/pyNN/models/neural_projections/connectors/fixed_number_post_connector.py

```diff
@@ -111,35 +201,71 @@ def create_synaptic_block(
+        lo = pre_vertex_slice.lo_atom
+        hi = pre_vertex_slice.hi_atom
+
+        # Get number of connections
+        n_connections = 0
+        for n in range(0, self._n_pre_neurons):
+            if (n >= lo and n <= hi):
+                n_connections += len(
+                    self._post_neurons_in_slice(post_vertex_slice, n))
+
+        # If self-connections are not allowed then subtract those connections
         if (not self._allow_self_connections and
                 pre_vertex_slice is post_vertex_slice):
-            n_connections -= len(post_neurons_in_slice)
+            for n in range(0, self._n_pre_neurons):
+                if (n >= lo and n <= hi):
+                    post_neurons = self._post_neurons_in_slice(
+                        post_vertex_slice, n)
+                    for m in range(0, len(post_neurons)):
+                        if (n == post_neurons[m]):
+                            n_connections -= 1
 
+        # Set up the block
         block = numpy.zeros(
             n_connections, dtype=AbstractConnector.NUMPY_SYNAPSES_DTYPE)
 
+        # If self connections not allowed set up source and target accordingly
         if (not self._allow_self_connections and
                 pre_vertex_slice is post_vertex_slice):
-            block["source"] = [
-                pre_index for pre_index in range(
-                    pre_vertex_slice.lo_atom, post_vertex_slice.hi_atom + 1)
-                for post_index in post_neurons_in_slice
-                if pre_index != post_index]
-            block["target"] = [
-                post_index for pre_index in range(
-                    pre_vertex_slice.lo_atom, post_vertex_slice.hi_atom + 1)
-                for post_index in post_neurons_in_slice
-                if pre_index != post_index]
+            pre_neurons_in_slice = []
+            post_neurons_in_slice = []
+            pre_vertex_array = numpy.arange(
+                pre_vertex_slice.lo_atom, pre_vertex_slice.hi_atom + 1)
+            for n in range(0, self._n_pre_neurons):
+                if (n >= lo and n <= hi):
+                    post_neurons = self._post_neurons_in_slice(
+                        post_vertex_slice, n)
+                    for m in range(0, len(post_neurons)):
+                        if (n != post_neurons[m]):
+                            post_neurons_in_slice.append(post_neurons[m])
+                            pre_neurons_in_slice.append(
+                                pre_vertex_array[n-pre_vertex_slice.lo_atom])
+
+            block["source"] = pre_neurons_in_slice
+            block["target"] = post_neurons_in_slice
         else:
-            block["source"] = numpy.repeat(numpy.arange(
-                pre_vertex_slice.lo_atom, pre_vertex_slice.hi_atom + 1),
-                len(post_neurons_in_slice))
-            block["target"] = numpy.tile(
-                post_neurons_in_slice, pre_vertex_slice.n_atoms)
+            pre_neurons_in_slice = []
```

Comment:
This "else" seems like it could be done by instead changing line 245 to say:
```
allow_self_connections = self._allow_self_connections or pre_vertex_slice is not post_vertex_slice
...
if n != post_neurons[m] or allow_self_connections:
```